### ISSUE TITLE
Distinct mTLS (client certificate) authentication for multiple paths on the same host

### ISSUE THREAD

**Issue by Okeanos** [traefik/traefik#9202]
### Welcome!

- [X] Yes, I've searched similar issues on [GitHub](https://github.com/traefik/traefik/issues) and didn't find any.
- [X] Yes, I've searched similar issues on the [Traefik community forum](https://community.traefik.io) and didn't find any.

### What did you expect to see?

As per the comment in #3575 I am opening a new issue to track this feature request.

What I need is the following:

The ability to specify mTLS (mutual TLS / client certificate) authentication with distinct CAs for different paths on the same host address. The ability to use `RequireAndVerifyClientCert` in its current form is insufficient because I want the different paths to be treated differently with separate CAs applying to each. "Overlapping" CAs, i.e. I don't know which of them is actually used for verification, are not a solution to my problem. I also cannot implement mTLS in the backend applications (which would render the mTLS option in Traefik obsolete anyway).

Related threads / discussions that I found that also require this or something similar:

- https://community.traefik.io/t/default-tls-options-without-host/13238
- https://community.traefik.io/t/many-found-different-tls-options-for-routers-on-the-same-host-errors/13852
- https://community.traefik.io/t/restrict-http-url-path-using-mtls/10650
- https://community.traefik.io/t/multiple-mtls-options-based-on-path-rules/4602
- #6362

None of these have an actual answer :(

---

Additional details/example:

- Given a host address such as `example.org` with an inability to assign additional TLDs or subdomains to the Traefik instance
- no reasonable way to control the TCP/UDP port
- the option to freely choose context paths below that host address such as:
  - `example.org/some-path/`
  - `example.org/different-path/`
- no useful way to implement mTLS in the backends proxied by Traefik (one per context path)

I want to apply different, distinct non-overlapping certificate authorities to the paths. That is, `some-path` is verified by `Some CA` and `different-path` is verified by `Different CA`. The CAs have no common ancestor by design. The currently available feature in Traefik [`RequireAndVerifyClientCert`](https://doc.traefik.io/traefik/https/tls/#client-authentication-mtls) is insufficient because it allows cross-verification, i.e. someone offering a `Some CA`-signed client certificate on `different-path` would be allowed through even though I want these requests to be verified by `Different CA` _only_ (and vice versa).

**Comment by ecki** [comment]
The server can request clients from multiple issuers in the protocol. But then the client needs to know which one applies (if the clients have keys for both they would randomly pick them). The server cannot decide to sent different issuer requests in the initial handshake because it does not know the route of the request at this time.

**Comment by Okeanos** [comment]
> The server can request clients from multiple issuers in the protocol. But then the client needs to know which one applies (if the clients have keys for both they would randomly pick them). The server cannot decide to sent different issuer requests in the initial handshake because it does not know the route of the request at this time.

So, if I understand correctly what you mean to say is:

TLS handshake negotiation as an RFC-standardized protocol only includes the hostname – which means distinguishing the certification path required by the client cannot be determined on HTTP only information (path, query parameters). This is because the path isn't actually send yet because the connection hasn't been established fully (as this would require lower level TLS to be finished d'oh); it's not a matter of making the TLS process aware of e.g. the path and including it at that time in the verification but it not being available at that point in time at all, by protocol design.

What should theoretically be possible but (for Traefik) not a good idea implementation-wise is to do a two-step verification:

1.  Regular TLS handshake accepting either certificate
2. Once HTTP level is reached extract the `path` information and _recheck_ the given certificate whether it matches the additional "only on _this_ `path` constraint" I outlined in my request?

My example here being basically implementing mTLS in the backend application (a second time) for more in depth checks.

Am I understanding you correctly?

**Comment by ecki** [comment]
Yes I was referring to CertificateRequest during handshake. The post handshake reauth might be possible, but that’s not very future proof it seems like tls1.3,h/2 and quic would not like it.

**Comment by Okeanos** [comment]
Thanks for confirming. Just another nail in the coffin of mutual TLS considering the state of e.g. [CORS mTLS handling](https://github.com/whatwg/fetch/issues/869) – granted the workaround in my problem is easy (separate domain!).

**Comment by rtribotte** [comment]
Hello @Okeanos,

According to your last message, it seems that we failed to resolve this issue in time!

I'm closing it now. Don't hesitate to ask to reopen it if you think it's necessary.
